test/distributed/_sharded_tensor/ops/test_linear.py

```diff
@@ -39,15 +46,15 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
         sharded_linear = torch.nn.Linear(*linear_size)
 
         # Copy the weights and bias from local linear
-        sharded_linear.weight = local_linear.weight
-        sharded_linear.bias = local_linear.bias
+        sharded_linear.weight = torch.nn.Parameter(local_linear.weight.detach().clone())
+        sharded_linear.bias = torch.nn.Parameter(local_linear.bias)
```

Comment:
> Great work! A few general comments:
> 
> 1. Lets move the nn/functional changes to a separate PR instead of clubbing it with this PR.
> 2. Lets make the nn/functional APIs consistent with the ones in distributed_c10d.py. Essentially you should have to just change one line and call the functional one and not need to change the existing implementation of sharded linear at all.

Sure, sounds good to me. I will separate them.